src/midgard/pointll.cc

```diff
@@ -124,77 +124,73 @@ float PointLL::Heading(const PointLL& ll2) const {
 // Finds the closest point to the supplied polyline as well as the distance
 // squared to that point and the index of the segment where the closest point
 // lies.
-std::tuple<PointLL, float, int>
-PointLL::ClosestPoint(const std::vector<PointLL>& pts, size_t begin_index, float dist_cutoff) const {
-  PointLL closest{};
-  int closest_segment = -1;
-  float mindistsqr = std::numeric_limits<float>::max();
-  size_t process_size = 0;
-
-  if (begin_index < pts.size()) {
-    process_size = (pts.size() - begin_index);
-  }
-
-  // If there are no points to process we are done
-  if (process_size == 0) {
-    return std::make_tuple(std::move(closest), std::move(mindistsqr), std::move(closest_segment));
-  }
+std::tuple<PointLL, float, int> PointLL::ClosestPoint(const std::vector<PointLL>& pts,
+                                                      int pivot_index,
+                                                      float forward_dist_cutoff,
+                                                      float reverse_dist_cutoff) const {
 
-  // If there is one point to process we are done
-  if (process_size == 1) {
-    return std::make_tuple(pts[begin_index], std::sqrt(DistanceSquared(pts[begin_index])),
-                           begin_index);
-  }
-
-  // Longitude (x) is scaled by the cos of the latitude so that distances are
-  // correct in lat,lon space
-  float lon_scale = cosf(lat() * kRadPerDeg);
+  // setup
+  if (pts.empty() || pivot_index < 0 || pivot_index > pts.size() - 1)
+    return std::make_tuple(PointLL(), std::numeric_limits<float>::max(), -1);
 
+  int closest_segment = pivot_index;
+  PointLL closest = pts[pivot_index];
   DistanceApproximator approx(*this);
-  PointLL point;
-  for (size_t index = begin_index; index < pts.size() - 1; ++index) {
-    // Get the current segment
-    const PointLL& u = pts[index];
-    const PointLL& v = pts[index + 1];
-
-    // Project a onto b where b is the origin vector representing this segment
-    // and a is the origin vector to the point we are projecting, (a.b/b.b)*b
-    auto bx = v.lng() - u.lng();
-    auto by = v.lat() - u.lat();
-
-    // Scale longitude when finding the projection. Only need the numerator
-    // at first, saving the division for the case where the lat,lon projects
-    // along the current segment.
-    auto bx2 = bx * lon_scale;
-    auto sq = bx2 * bx2 + by * by;
-    auto scale = (lng() - u.lng()) * lon_scale * bx2 + (lat() - u.lat()) * by;
-
-    // Projects along the ray before u
-    if (scale <= 0.f) {
-      point = {u.lng(), u.lat()};
-    } // Projects along the ray after v
-    else if (scale >= sq) {
-      point = {v.lng(), v.lat()};
-    } // Projects along the ray between u and v
-    else {
-      scale /= sq;
-      point = {u.lng() + bx * scale, u.lat() + by * scale};
-    }
+  float mindistsqr = approx.DistanceSquared(closest);
+
+  // start going backwards, then go forwards
+  for (int reverse = 0; reverse < 2; ++reverse) {
```

Comment:
i could make tuples out of the 3 things below and iterate over a vector of 2 tuples. this might be more performant but is probably in the noise

Reply:
I think the ternary expressions below are OK, they're pretty self contained.

*But*, I think it would be a bit more readable to use this form of the loop and make `reverse` a boolean:

`for (bool reverse : { true, false })`

Reply:
haven't seen this before but I like it